python/benchmark/gen_data_distributed.py

```diff
@@ -695,26 +822,57 @@ def make_sparse_regression_udf(
                 generator_p = np.random.RandomState(partition_seeds[partition_index])
 
             # Label Calculation
-            y = sparse_matrix.dot(ground_truth) + bias
+            if multinomial_log:
+                y = np.squeeze(
+                    np.array(
+                        [
+                            [sparse_matrix.dot(class_truth) + bias]
+                            for class_truth in ground_truth
+                        ]
+                    )
+                )
+                y = np.transpose(y)
+            else:
+                y = sparse_matrix.dot(ground_truth) + bias
 
             # Type conversion
             if use_cupy:
-                y = cp.asarray(y)
+                y_p = cp.asarray(y)
             else:
-                y = np.asarray(y)
+                y_p = np.asarray(y)
 
             # Random Noise
             if noise > 0.0:
-                y += generator_p.normal(scale=noise, size=y.shape)
+                y_p += generator_p.normal(scale=noise, size=y.shape)
 
             # Logistric Regression sigmoid and sample
             if logistic_regression:
-                if use_cupy:
-                    prob = 1 / (1 + cp.exp(-y))
-                    y = cp.random.binomial(1, prob)
+                if multinomial_log:
+                    if use_cupy:
+                        y = y_p.get()
+                        del y_p
+
+                    probs = [sp.special.softmax(target_weight) for target_weight in y]
+
+                    multi_labels = [
+                        random.choices(range(n_classes), weights=p)[0] for p in probs
+                    ]
+
+                    if use_cupy:
+                        y = cp.asarray(multi_labels)
+                    else:
+                        y = np.asarray(multi_labels)
                 else:
-                    prob = 1 / (1 + np.exp(-y))
-                    y = np.random.binomial(1, prob)
+                    if use_cupy:
+                        prob = 1 - 1 / (1 + cp.exp(-y_p))
+                        del y_p
+                        y = cp.random.binomial(1, prob)
+                    else:
+                        prob = 1 - 1 / (1 + np.exp(-y_p))
+                        del y_p
+                        y = np.random.binomial(1, prob)
+            else:
```

Comment:
large chunk of code looks identical in dense class and sparse class. Can we merge identical code (e.g. create a python function)?

Reply:
Seems like an extra user defined function inside the UDF would break the pyspark dependency system. I get the same "Unable to find module gen_data_distribtued" error as last time.